NoticeController tests

charliescheer opened the pull request:
### Fix
This PR is a beginning attempt at adding unit tests for the NoticeController class.  This makes a few changes to access in the NoticeController class and writes some basic unit tests for the functionality of the controller methods.

One thing that I particularly could use some advice on is the internal init that I created so I can override the NoticePresenter with a mock.  Since the NoticeController is a singleton, I feel like it is a bad idea to have an initializer that is exposed, but I couldn't find another way around that.  Would be happy to find a way to be able to override the presenter without exposing an initializer to the controller.

### Test
***(Required)*** List the steps to test the behavior.  For example:
1. look over the access changes and make sure that  I am not exposing too much
2. run the tests to ensure they all pass


### Review
***(Required)*** Add instructions for reviewers.  For example:
> Only one develope is required to review these changes, but anyone can perform the review.

### Release
> These changes do not require release notes.

charliescheer committed:
Add controller test file to project

charliescheer committed:
Updated NoticePresenter to conform to NoticePresentable protocol

charliescheer committed:
Updated NoticePresentingDelegate to NoticeInteractionDelegate

charliescheer committed:
Exposed access to some variable values for testing purposes

charliescheer committed:
Added unit tests for NoticeController

eshurakov commented on Simplenote/NoticeController.swift:
```diff
@@ -7,18 +7,24 @@ class NoticeController {
     private var timer: Timer?
 
-    private var isPresenting: Bool {
+    internal var isPresenting: Bool {
```
You don't need to specify `internal` as it should be the default visibility.

eshurakov commented on Simplenote/NoticeController.swift:
```diff
@@ -7,18 +7,24 @@ class NoticeController {
     private var notices: [Notice] = []
     private var current: Notice?
-    private let noticePresenter = NoticePresenter()
+    private var noticePresenter: NoticePresentable
```
This can stay as `let` :-)

eshurakov commented on Simplenote/NoticeController.swift:
```diff
@@ -7,18 +7,24 @@ class NoticeController {
+
+    internal init(presenter: NoticePresentable) {
+        self.noticePresenter = presenter
+    }
```
Can be `init(presenter: NoticePresentable = NoticePresenter())`

eshurakov commented on Simplenote/NoticePresenter.swift:
```diff
@@ -1,6 +1,12 @@
+    func startListeningToKeyboardNotifications()
+    func presentNoticeView(_ noticeView: NoticeView, completion: @escaping (Bool) -> Void)
+    func dismissNotification(completion: @escaping () -> Void)
+}
```
You don't necessarily need a protocol to mock `NoticePresenter`, you can also subclass it and override exposed methods.

eshurakov commented on SimplenoteTests/NoticeContorllerTests.swift:
```diff
@@ -1,33 +1,142 @@
+    controller.setupNoticeController()
+
+    XCTAssertTrue(presenter.listeningToKeyboard == true)
+    }
```
Alignment looks off :-)

eshurakov commented on SimplenoteTests/NoticeContorllerTests.swift:
```diff
@@ -1,33 +1,142 @@
 
-    override func tearDownWithError() throws {
-        // Put teardown code here. This method is called after the invocation of each test method in the class.
+    func testControllerPresent() {
+        let expectation = XCTestExpectation(description: "Did dismiss notice")
+        let timer = Timer.scheduledTimer(withTimeInterval: TimeInterval(0.5), repeats: false) { (_) in
+            expectation.fulfill()
+        }
+
+        let notice = Notice(message: "Message", action: nil)
+        controller.present(notice, withTimer: timer)
+
+        wait(for: [expectation], timeout: TimeInterval(3))
+
     }
```
It is best to avoid using delays in unit tests as they should be fast. You can mock a timer and "fire" it when needed.
It's also not clear what is being tested here. That timer would fire? It would fire even if it's not passed to controller :-)

charliescheer replied:
Yeah, I was testing that the timer in the controller present method fired.  I have removed that test

eshurakov commented on SimplenoteTests/NoticeContorllerTests.swift:
```diff
@@ -1,33 +1,142 @@
+
+        XCTAssertTrue(presenter.presented)
+
+        let presentedNoticeView = try XCTUnwrap(presenter.noticeView)
+        XCTAssertEqual(presentedNoticeView.message, notice.message)
+        XCTAssertTrue(presentedNoticeView.handler == nil)
```
This is the same as in `testMakeNoticeView`. Because they're testing the same thing, we can remove checks from this method.

eshurakov commented on SimplenoteTests/NoticeContorllerTests.swift:
```diff
@@ -1,33 +1,142 @@
-            // Put the code you want to measure the time of here.
+    func testMakeNoticeViewWithAction() {
+        let action = NoticeAction(title: "Title") {
+            print("Action")
```
Leftover debug code? :-)

eshurakov commented on SimplenoteTests/NoticeContorllerTests.swift:
```diff
@@ -1,33 +1,142 @@
-        // This is an example of a performance test case.
-        self.measure {
-            // Put the code you want to measure the time of here.
+    func testMakeNoticeViewWithAction() {
+        let action = NoticeAction(title: "Title") {
+            print("Action")
         }
+        let notice = Notice(message: "Message ", action: action)
+
+        controller.present(notice)
+        XCTAssertEqual(notice.message, presenter.noticeView?.message)
+        XCTAssertEqual(notice.action?.title, presenter.noticeView?.actionTitle)
+        XCTAssertNotNil(presenter.noticeView?.handler)
+    }
+
+    func testMakeNoticeViewEmptyNotice() {
+        let action = NoticeAction(title: "") {
+            print("")
+        }
+        let notice = Notice(message: "", action: action)
+
+        controller.present(notice)
+        XCTAssertEqual(notice.message, presenter.noticeView?.message)
+        XCTAssertNotEqual(notice.action?.title, presenter.noticeView?.actionTitle)
+        XCTAssertNotNil(presenter.noticeView?.handler)
     }
```
`testMakeNoticeViewWithAction` and `testMakeNoticeViewEmptyNotice` are the same except the message. Maybe we can extract a common method?

eshurakov commented on SimplenoteTests/NoticeContorllerTests.swift:
```diff
@@ -1,33 +1,142 @@
+        XCTAssertFalse(controller.isPresenting)
+    }
+
+    func testDismissContinuesToNextNotice() {
+        let noticeA = Notice(message: "Message A", action: nil)
+        let noticeB = Notice(message: "Message B", action: nil)
+        controller.present(noticeA, withTimer: NoticeContorllerTests.noActionTimer)
+        controller.present(noticeB, withTimer: NoticeContorllerTests.noActionTimer)
+
+        XCTAssertEqual(controller.pendingNotices, 1)
+
+        controller.dismiss()
+
+        XCTAssertTrue(presenter.dismissed)
+        XCTAssertTrue(controller.isPresenting)
+        XCTAssertEqual(controller.pendingNotices, 0)
+    }
```
My understanding is that we want to test here that notices are presented one after the other. The best way to do that is to have a log of events in `MockNoticePresenter`. 
For example presenting 2 notices should result in the following log: `present(noticeA), dismiss(noticeA), present(noticeB), dismiss(noticeB)`. Later in the test we can compare the log with what we expect. 
I think this way the behaviour is more deterministic.

charliescheer replied:
Okay @eshurakov I have updated the tests as well.  I left a few questions for you in slack, so not all of these comments have been handled, but they are ready for another look.  Thanks!!!

eshurakov reviewed (commented):
Hey Charlie, good work on tests.
Sending you some comments.

charliescheer committed:
Updated controller unit tests

charliescheer committed:
Merge branch 'feature/in-app-notifications-controller' into feature/in-app-notifications-controller-tests

charliescheer committed:
Updated noticeContorllerTests to use time factory

eshurakov commented on Simplenote/NoticeController.swift:
```diff
@@ -7,7 +7,7 @@ class NoticeController {
     private var notices: [Notice] = []
     private var current: Notice?
-    private let noticePresenter = NoticePresenter()
+    private var noticePresenter: NoticePresenter
```
This can stay `let` because it is assigned in the init :-)

eshurakov replied:
🤔

eshurakov commented on Simplenote/NoticeController.swift:
```diff
@@ -16,14 +16,20 @@ class NoticeController {
+    init(presenter: NoticePresenter, timerFactor: TimerFactory) {
+        self.noticePresenter = presenter
+        self.timerFactory = timerFactor
     }
```
There is no need to have 2 methods here :-)
Making the first one non-private would do the job. It's up to the caller whether to use default values or provide another value.
For example, this is a valid call `NoticeController(timerFactory: MockFactory())` that would use "default" presenter.

eshurakov commented on Simplenote/NoticePresenter.swift:
```diff
@@ -4,7 +4,7 @@ class NoticePresenter {
     // MARK: Properties
     //
-    private var noticeView: UIView?
+    var noticeView: UIView?
```
Why do we need to expose this? 🤔

eshurakov commented on SimplenoteTests/NoticeContorllerTests.swift:
```diff
@@ -1,33 +1,171 @@
+    }
+}
+
+class MockTimerFactory: TimerFactory {
+    var timer: Timer?
+
+    override func scheduledTimer(with timeInterval: TimeInterval, completion: @escaping () -> Void) -> Timer {
+        if let timer = timer {
+            timer.fire()
+
+            return timer
+        } else {
+            return Timer.scheduledTimer(withTimeInterval: NoticeContorllerTests.timeInterval, repeats: false, block: NoticeContorllerTests.timerNoActionCompletionHandler)
+        }
+    }
 }
```
What you want to do here is to return a mock timer that won't fire till you tell it to. This will allow you to precisely control test scenarios. Mock timer could look like this:
```
class MockTimerFactory: TimerFactory {
    var timer: Timer?

    override func scheduledTimer(with timeInterval: TimeInterval, completion: @escaping () -> Void) -> Timer {
        let timer = MockTimer()
        timer.completion = completion
        return timer
    }
}

class MockTimer: Timer {
    var completion: (() -> Void)?

    override func fire() {
        completion?()
    }

    override func invalidate() {
        super.invalidate()
        completion = nil
    }
}
```

eshurakov commented on SimplenoteTests/NoticeContorllerTests.swift:
```diff
@@ -1,33 +1,171 @@
+    }
+}
+
+class MockNoticePresenter: NoticePresenter {
+    var presented: Bool = false
+    var dismissed: Bool = false
+    var listeningToKeyboard: Bool = false
+    var presentedViews = [NoticeView]()
+    var actionLog = String()
+
+    override func presentNoticeView(_ noticeView: NoticeView, completion: @escaping () -> Void) {
+        self.noticeView = noticeView
+        presentedViews.append(noticeView)
+        actionLog.append("presentNoticeView(\(noticeView.message ?? "")), ")
+        presented = true
+
+        completion()
+    }
+
+    override func dismissNotification(completion: @escaping () -> Void) {
+        dismissed = true
+        actionLog.append("dismissNotification(), ")
+        completion()
+    }
+
+    override func startListeningToKeyboardNotifications() {
+        listeningToKeyboard = true
+    }
+}
```
If we use an action log there is no reason to keep also `presented` and `dismissed`. And instead of a list of presented views we can keep only the last one.
Mock presenter could look like this:
```
class MockNoticePresenter: NoticePresenter {
    enum Action: Equatable {
        case present(String?)
        case dismiss(String?)
    }

    var actionLog: [Action] = []

    var listeningToKeyboard: Bool = false
    var lastNoticeView: NoticeView?

    override func presentNoticeView(_ noticeView: NoticeView, completion: @escaping () -> Void) {
        lastNoticeView = noticeView
        actionLog.append(.present(noticeView.message))
        completion()
    }

    override func dismissNotification(completion: @escaping () -> Void) {
        let noticeView = try! XCTUnwrap(lastNoticeView)
        lastNoticeView = nil
        actionLog.append(.dismiss(noticeView.message))
        completion()
    }

    override func startListeningToKeyboardNotifications() {
        listeningToKeyboard = true
    }
}
```

eshurakov commented on SimplenoteTests/NoticeContorllerTests.swift:
```diff
@@ -1,33 +1,171 @@
+        XCTAssertFalse(controller.isPresenting)
+    }
+
+    func testDismissContinuesToNextNotice() {
+        let noticeA = Notice(message: "Message A", action: nil)
+        let noticeB = Notice(message: "Message B", action: nil)
+        controller.present(noticeA)
+        controller.present(noticeB)
+
+        XCTAssertTrue(controller.isPresenting)
+        XCTAssertEqual(controller.pendingNotices, 1)
+
+        controller.dismiss()
+
+        XCTAssertTrue(controller.isPresenting)
+        XCTAssertEqual(controller.pendingNotices, 0)
+
+        controller.dismiss()
+
+        XCTAssertTrue(presenter.dismissed)
+        XCTAssertFalse(controller.isPresenting)
+        XCTAssertEqual(controller.pendingNotices, 0)
+        XCTAssertEqual(presenter.actionLog, "presentNoticeView(Message A), dismissNotification(), presentNoticeView(Message B), dismissNotification(), ")
+    }
```
An example how actionLog and mock timer can be used (I didn't run the actual test :-) )
```
func testDismissContinuesToNextNotice() {
    let noticeA = Notice(message: "Message A", action: nil)
    let noticeB = Notice(message: "Message B", action: nil)
    controller.present(noticeA)
    controller.present(noticeB)

    var expectedActions: [MockNoticePresenter.Action] = [
        .present(noticeA.message)
    ]
    XCTAssertEqual(presenter.actionLog, expectedActions)

    timerFactory.timer?.fire()

    expectedActions.append(.dismiss(noticeA.message))
    expectedActions.append(.present(noticeB.message))
    XCTAssertEqual(presenter.actionLog, expectedActions)

    timerFactory.timer?.fire()

    expectedActions.append(.dismiss(noticeB.message))
    XCTAssertEqual(presenter.actionLog, expectedActions)
}
```

This way we mimic how the actual flow would work as the client code wouldn't call `dismiss`.

We also don't need `controller.isPresenting` and `controller.pendingNotices`. And `controller.dismiss` can remain private.

eshurakov reviewed (commented):
Some other things we can test:
* pressing on action button dismisses the notice
* long press invalidates the timer and notice stays on screen
* when long press is released, timer is created and notice is dismissed after that

jleandroperez commented on SimplenoteTests/NoticeContorllerTests.swift:
```diff
@@ -1,33 +1,171 @@
 import XCTest
+@testable import Simplenote
 
 class NoticeContorllerTests: XCTestCase {
```
Nipticky: typo in the classname (Controller / Contorller)

charliescheer committed:
Merge branch 'feature/in-app-notifications-controller' into feature/in-app-notifications-controller-tests

charliescheer committed:
Removed duplicate init method for NoticeController

charliescheer committed:
Created MockTimer class for greater control of the timers act in test

charliescheer committed:
fixed typo in NoticeControllerTests class

charliescheer committed:
Changed MockNoticePresenter to maintain an action log for testing action

charliescheer committed:
Updated mockTimer to take a completion handler at init

charliescheer committed:
Updated mockTimerFactory to store a mockTimer

charliescheer committed:
Updated MockTimerFactory to store a mock timer

charliescheer committed:
Added tests for NoticeController - NoticeInteractionDelegate methods

charliescheer commented:
Thanks for all of the suggestions!!  The action log is a great idea.  Will put that idea in my testing tool kit :)

I have updated the tests and added some new ones.  Let me know if there are any more notes @eshurakov !!

eshurakov commented on SimplenoteTests/NoticeControllerTests.swift:
```diff
@@ -0,0 +1,238 @@
+        XCTAssertNil(timerFactory.timer?.completion)
+    }
+
+    func testLongPressReleasedDismissesTimer() {
+        let noticeA = Notice(message: "Message A", action: nil)
+
+        controller.present(noticeA)
+        var expectedActions: [MockNoticePresenter.Action] = [
+            .present(noticeA.message)
+        ]
+
+        controller.noticePressBegan()
+
+        XCTAssertEqual(expectedActions, presenter.actionLog)
+        XCTAssertNil(timerFactory.timer?.completion)
+
+        controller.noticePressEnded()
+        timerFactory.timer?.fire()
+        expectedActions.append(.dismiss(noticeA.message))
+
+        XCTAssertEqual(expectedActions, presenter.actionLog)
+    }
```
Just a note: because we already tested some parts in other places, we can only assert things that we didn't check yet.
For example, because most of this method is tested in `testLongPressInvalidatesTimerAndNoticeDoesNotDismiss` it could look like:
```
func testLongPressReleasedDismissesTimer() {
    let noticeA = Notice(message: "Message A", action: nil)

    controller.present(noticeA)
    controller.noticePressBegan()
    controller.noticePressEnded()

    XCTUnwrap(timerFactory.timer).fire() <- can use XCTUnwrap to make sure we actually created a timer

    var expectedActions: [MockNoticePresenter.Action] = [
      .present(noticeA.message),
      .dismiss(noticeA.message)
    ]
    XCTAssertEqual(expectedActions, presenter.actionLog)
}
```

eshurakov commented on SimplenoteTests/NoticeControllerTests.swift:
```diff
@@ -0,0 +1,238 @@
+        controller.present(notice)
+
+        XCTAssertEqual(notice.message, presenter.lastNoticeView?.message)
+        XCTAssertNotEqual(notice.action?.title, presenter.lastNoticeView?.actionTitle)
```
To continue the previous conversation, we can update actionTitle getter in NoticeView to `noticeButton.title(for: .normal)`. That would make it behave the same as `testMakeNoticeViewWithAction`

eshurakov commented on SimplenoteTests/NoticeControllerTests.swift:
```diff
@@ -0,0 +1,238 @@
+        XCTAssertEqual(presenter.lastNoticeView?.message, noticeA.message)
+    }
+
+    func testAppendToQueueIfNew() {
+        let noticeA = Notice(message: "Message A", action: nil)
+        let noticeB = Notice(message: "Message B", action: nil)
+
+        controller.present(noticeA)
+        XCTAssertEqual(controller.pendingNotices, 0)
+
+        controller.present(noticeA)
+        XCTAssertEqual(controller.pendingNotices, 0)
+
+        controller.present(noticeB)
+        XCTAssertEqual(controller.pendingNotices, 1)
+
+        controller.present(noticeB)
+        XCTAssertEqual(controller.pendingNotices, 1)
+    }
```
Do you think we can also use actionLog for this and get rid of `pendingNotices`?

charliescheer replied:
Had to think about that one for a little bit, but yes I think we can!

eshurakov commented on Simplenote/NoticeController.swift:
```diff
@@ -16,14 +16,15 @@ class NoticeController {
     }
 
-    private var isPresenting: Bool {
+    var isPresenting: Bool {
```
This can stay private :-)

charliescheer replied:
I merged some changes from the branch that I think cause some properties to change in ways I had not intended....  Fixed

eshurakov reviewed (commented):
Thank you @charliescheer, looking better and better.

Sending you some more comments :-)

charliescheer committed:
Updated timers in tests to be unwrapped before firing

charliescheer committed:
Simplified test testLongPressReleasedDismissesTimer

charliescheer committed:
refactored makeNoticeTests to minimize repeat code

charliescheer committed:
Updated testAppendToQueueIfNew to use action log instead of counts

charliescheer committed:
Updated some property definitions in notice controller

charliescheer commented:
@eshurakov thanks again for these additional suggestions.  I have updated the tests per your suggestions.

eshurakov commented on SimplenoteTests/NoticeControllerTests.swift:
```diff
@@ -37,21 +37,31 @@ class NoticeControllerTests: XCTestCase {
         controller.present(noticeB)
-        XCTAssertEqual(controller.pendingNotices, 1)
+
+        var expectedActions: [MockNoticePresenter.Action] = [
+            .present(noticeA.message)
+        ]
+        XCTAssertEqual(expectedActions, presenter.actionLog)
+
+        try XCTUnwrap(timerFactory.timer).fire()
+        expectedActions.append(.dismiss(noticeA.message))
+        expectedActions.append(.present(noticeB.message))
+
+        XCTAssertEqual(expectedActions, presenter.actionLog)
+
+        try XCTUnwrap(timerFactory.timer).fire()
+        expectedActions.append(.dismiss(noticeB.message))
+
+        XCTAssertEqual(expectedActions, presenter.actionLog)
```
Awesome!

eshurakov reviewed (approved):
Thank you for adding tests!

Looks good, let's :shipit:
